### PR TITLE
Small refactorings

### PR DESCRIPTION
**Pull request by ValentinAebi**
While writing the report I noticed some bugs and some things that could be improved. This PR contains the modifications that I made following these observations.

#### Checkers
* I noticed that people who define their own checkers could want to report several violations on a single node for the same checker, so I changed the return value of `checkNode` from `Option[Violation]` to `List[Violation]`. To keep it easy to use I defined a method in `Violation` that converts a `Violation` to a list containing only this `Violation`: `violation.toSingletonList`
* In some cases, the fact that `BlacklistRules` and `Features` require a `PartialFunction` to be passed as a constructor argument causes compilation errors (mentioning some premature access). So I modified the `PartialFunction` into a `val` that the `BlacklistRule` or `Feature` must implement.
* I modified the `BlacklistChecker` so that it reports several violations on the same node if the node violates several rules (IIRC @julienrf asked me to do it but I had forgotten...)

#### Code transformation
* I noticed several bugs that I fixed. One of them was that I had forgotten to consider `do-while` loops in the transformation (so they were simply left unchanged)...
* I added more tests. The code of 5 tests is taken from exercises from the 2020 final exam from EPFL's functional programming class (I cited the source in the tests).
* The transformation system now simply rejects programs containing `return`s

### CODE REVIEW

**Comment by civp** [comment]
Great work! Then we will need to modify the testing infrastructure to allow multiple violations on the same node.

**Comment by ValentinAebi** [comment]
As discussed with @civp this PR also removes most of the tests in `syntactic/src/test`, as they are redundant with the ones of the testing infrastructure. I keep only one test that checks the content of `PredefFeatures.allDefinedfeatures`.

File: functional-converter/src/main/scala/translator/RestrictionsEnforcer.scala [hunk 1]
```diff
@@ -95,7 +113,27 @@ class RestrictionsEnforcer(reporter: Reporter) {
     }.flatten
 
     // use map and then exists to avoid stopping at the first conflict because they must be written in the reporter
-    !allPairs(definedNames).map(names => checkConflict(names._1, names._2)).exists(identity)
+    val conflictsCheck = !allPairs(definedNames).map(names => checkConflict(names._1, names._2)).exists(identity)
```
Comment (julienrf) [inline functional-converter/src/main/scala/translator/RestrictionsEnforcer.scala:116]: It seems that you could replace `.map(f).exists(identity)` with `.exists(f)`

File: syntactic/src/main/scala/syntactic/blacklist/BlacklistChecker.scala [hunk 2]
```diff
@@ -11,17 +11,14 @@ class BlacklistChecker private(rules: List[BlacklistRule]) extends Checker {
...
-  private val combinedCheckFunc: PartialFunction[Tree, Option[Violation]] = {
-    val checkFuncs = rules.map(_.checkFunc.andThen(Some(_)))
-    checkFuncs.reduceLeft(_.orElse(_))
-  }
+  private val checkFuncs = rules.map(_.checkFunc.orElse(defaultPartFunc))
 
-  override def checkNode(node: Tree): Option[Violation] = {
-    combinedCheckFunc.applyOrElse(node, defaultPartFunc)
+  override def checkNode(node: Tree): List[Violation] = {
+    checkFuncs.map(_.apply(node)).foldLeft(List.empty[Violation])(_ ++ _)
```
Comment (julienrf) [inline syntactic/src/main/scala/syntactic/blacklist/BlacklistChecker.scala:21]: It seems you can simplify `.map(f).foldLeft(Nil)(_ ++ _)` to `.flatMap(f)`.

**Review by julienrf — APPROVED** [review]
Nice work Valentin!

File: functional-converter/src/main/scala/translator/RestrictionsEnforcer.scala (hunk 1 above)
Comment (julienrf) [reply functional-converter/src/main/scala/translator/RestrictionsEnforcer.scala:116]: Ah, indeed if `f` is side-effecting my proposal does not apply :) Thank you for clarifying!

**Pull request merged**